src/sentry/static/sentry/app/views/settings/components/forms/tableField.tsx

```diff
@@ -0,0 +1,172 @@
+import {Flex, Box} from 'grid-emotion';
+import PropTypes from 'prop-types';
+import React from 'react';
+import styled from 'react-emotion';
+
+import {defined, objectIsEmpty} from 'app/utils';
+import {t} from 'app/locale';
+import Button from 'app/components/button';
+import InputField from 'app/views/settings/components/forms/inputField';
+import Input from 'app/views/settings/components/forms/controls/input';
+
+type Props = {
+  name?: string;
+  addButtonText?: React.ReactNode;
+  columnLabels: any;
+  columnKeys?: string[];
+  allowEmpty?: boolean;
+};
+
+export default class TableField extends React.Component<Props> {
+  static propTypes = {
+    ...InputField.propTypes,
+    /**
+     * Text used for the 'add' button. An empty string can be used
+     * to just render the "+" icon.
+     */
+    addButtonText: PropTypes.node,
+    /**
+     * A list of column labels (headers) for the table.
+     */
+    columnLabels: PropTypes.objectOf(PropTypes.node).isRequired,
+    /**
+     * A list of column keys for the table, in the order that you want
+     * the columns to appear - order doesn't matter in columnLabels
+     */
+    columnKeys: PropTypes.arrayOf(PropTypes.string),
+    /**
+     * Automatically save even if fields are empty
+     */
+    allowEmpty: PropTypes.bool,
+  };
+
+  static defaultProps = {
```

Comment:
@MeredithAnya I recommend using the pattern of using the types of `defaultProps` to set the type of your `props`. See `/components/events/eventDataSection.tsx` as an example.